compiler/noirc_evaluator/src/ssa/opt/array_set.rs

```diff
@@ -206,135 +206,47 @@ fn make_mutable(
-        builder.set_runtime(RuntimeType::Brillig(InlineType::default()));
-
-        let array_type = Type::Array(Arc::new(vec![Type::field()]), 5);
-        let zero = builder.field_constant(0u128);
-        let array_constant =
-            builder.array_constant(vector![zero, zero, zero, zero, zero], array_type.clone());
-        let nested_array_type = Type::Array(Arc::new(vec![array_type.clone()]), 2);
-        let nested_array_constant = builder
-            .array_constant(vector![array_constant, array_constant], nested_array_type.clone());
-
-        let v3 = builder.insert_allocate(array_type.clone());
-
-        builder.insert_store(v3, nested_array_constant);
-
-        let v4 = builder.insert_allocate(array_type.clone());
-        builder.insert_store(v4, nested_array_constant);
-
-        let b1 = builder.insert_block();
-        let zero_u32 = builder.numeric_constant(0u128, Type::unsigned(32));
-        builder.terminate_with_jmp(b1, vec![zero_u32]);
-
-        // Loop header
-        builder.switch_to_block(b1);
-        let v5 = builder.add_block_parameter(b1, Type::unsigned(32));
-        let five = builder.numeric_constant(5u128, Type::unsigned(32));
-        let v8 = builder.insert_binary(v5, BinaryOp::Lt, five);
-
-        let b2 = builder.insert_block();
-        let b3 = builder.insert_block();
-        let b4 = builder.insert_block();
-        let b5 = builder.insert_block();
-        builder.terminate_with_jmpif(v8, b3, b2);
-
-        // Loop body
-        // b3 is the if statement conditional
-        builder.switch_to_block(b3);
-        let two = builder.numeric_constant(5u128, Type::unsigned(32));
-        let v9 = builder.insert_binary(v5, BinaryOp::Eq, two);
-        builder.terminate_with_jmpif(v9, b4, b5);
-
-        // b4 is the rest of the loop after the if statement
-        builder.switch_to_block(b4);
-        let v10 = builder.insert_load(v3, nested_array_type.clone());
-        builder.insert_store(v4, v10);
-        builder.terminate_with_jmp(b5, vec![]);
-
-        builder.switch_to_block(b5);
-        let v11 = builder.insert_load(v3, nested_array_type.clone());
-        let twenty = builder.field_constant(20u128);
-        let v13 = builder.insert_array_get(v11, zero, array_type.clone());
-        let v14 = builder.insert_array_set(v13, v5, twenty);
-        let v15 = builder.insert_array_set(v11, v5, v14);
-
-        builder.insert_store(v3, v15);
-        let one = builder.numeric_constant(1u128, Type::unsigned(32));
-        let v17 = builder.insert_binary(v5, BinaryOp::Add, one);
-        builder.terminate_with_jmp(b1, vec![v17]);
-
-        builder.switch_to_block(b2);
-        builder.terminate_with_return(vec![]);
+        let src = "
+            brillig(inline) fn main f0 {
+              b0():
+                v1 = allocate -> [Field; 5]
+                store [[Field 0, Field 0, Field 0, Field 0, Field 0] of Field, [Field 0, Field 0, Field 0, Field 0, Field 0] of Field] of [Field; 5] at v1
```

Comment:
How do we interpret this? I thought `v1` would be an array of 5 fields, but it looks like `[[Field; 5]; 2]` 🤔

Reply:
Ah, right, I was also a bit confused by this.

The type after `of` is the type of the elements, not the type of the array itself.

Reply:
Great, TIL that allocate works on references, and I see the original version also does the same thing: `let v3 = builder.insert_allocate(array_type.clone());`, rather than `.insert_allocate(nested_array_type...`

Maybe some package level example would be helpful to illustrate and emphasise this syntax, but now that you spelled it out, it makes sense, thanks!